AmplifyPlugins/DataStore/Sources/AWSDataStorePlugin/Sync/SubscriptionSync/ReconcileAndLocalSave/RemoteSyncReconciler.swift

```diff
@@ -42,18 +42,35 @@ struct RemoteSyncReconciler {
     /// - Returns: disposition of models to apply locally
     static func getDispositions(_ remoteModels: [RemoteModel],
                                 localMetadatas: [LocalMetadata]) -> [Disposition] {
-        guard !remoteModels.isEmpty else {
-            return []
-        }
-
-        guard !localMetadatas.isEmpty else {
-            return remoteModels.compactMap { getDisposition($0, localMetadata: nil) }
```

Comment:
This line was causing the bug:

here we were geting an empty localMetadatas and remoteModels with two elements for the same model (the same ID):
eg. [
    ModelA - id: aaa, version: 1, deleted: false,
    ModelA - id: aaa, version: 2, deleted: true
]

and then "deteled true disposition" was dropped so only "create disposition" was returned

Reply:
Let me try to summarize this, the reconciliation path for DataStore takes the sync query remote model results and decides to apply the remote model to the local database or drop it. Previously the logic processed one model at a time. We changed this to a batch operation by putting the reconciliation logic under a transaction block, operating on the entire list of remote models. The necessary logic updates to perform the SQL operations more efficiently included constructing a predicate of all the model identifiers as a filter criteria for querying local data that is used for the reconciliation logic. There exists a bug with this optimization when having multiple entries of remote model belong to the same identifier and the changes to the local model ends up as a deleted model. 

Bug example
remoteModels = `[id=111 version=1 deleted=false, id=111 version=2 deleted=true]`

When the disposition of the model is calculated for the create model, it is to create the model with the local database, because there is no local model, nor is there a local model with version less than the remote model.
When the disposition of the model is calculated for the delete model, it is to drop it because no local model yet to exist, so it thinks there is nothing to delete.

The end result is the model is created but not deleted. The fix we are applying in the PR is to optimize the reconciliation by only taking the latest versions of the remote model. In the example, this becomes `[id=111 version=2, deleted=true]`. The latest version represents the final state of the model that we want applied to the local database. In this case, the remote model will be dropped. There is no create disposition so it will not be created either.